pricegraph/src/api/price_estimation.rs

```diff
@@ -207,12 +235,197 @@ mod tests {
+            0.5 / FEE_FACTOR
+        );
+
+        // NOTE: At this point, it is worth it to start using the liquidity
+        // from the second 3->4 order at the `100:1` limit price.
+        assert_approx_eq!(
+            pricegraph
+                .estimate_limit_price(TokenPair { buy: 4, sell: 3 }, 101_000_000.0 * FEE_FACTOR)
+                .unwrap(),
+            0.01 / FEE_FACTOR.powi(2)
+        );
+        assert_approx_eq!(
+            pricegraph
+                .estimate_limit_price(TokenPair { buy: 4, sell: 3 }, 200_000_000.0 * FEE_FACTOR)
+                .unwrap(),
+            0.01 / FEE_FACTOR.powi(2)
+        );
+
+        // NOTE: If we buy all available token 4 then would receive at most
+        // `2_000_000 / FEE_FACTOR`.
+        assert_approx_eq!(
+            pricegraph
+                .estimate_limit_price(TokenPair { buy: 4, sell: 3 }, 400_000_000.0)
+                .unwrap(),
+            0.005 / FEE_FACTOR
+        );
+    }
+
+    #[test]
+    fn estimated_buy_amount_monotonically_increasing() {
+        //    /-0.04--v
+        //   /--0.02--v
+        //  /---0.01--v
+        // 1           2
+        let pricegraph = pricegraph! {
+            users {
+                @1 {
+                    token 2 => 1_000_000_000,
+                }
+            }
+            orders {
+                owner @1 buying 1 [1_000_000] selling 2 [100_000_000],
+                owner @1 buying 1 [2_000_000] selling 2 [100_000_000],
+                owner @1 buying 1 [4_000_000] selling 2 [100_000_000],
+            }
+        };
+
+        // NOTE: Partially use liquidity provided by the first order.
+        assert_approx_eq!(
+            pricegraph
+                .order_for_sell_amount(TokenPair { buy: 2, sell: 1 }, 500_000.0)
+                .unwrap()
+                .buy,
+            50_000_000.0 / FEE_FACTOR.powi(2)
+        );
 
-        let price = pricegraph.estimate_limit_price(TokenPair { buy: 2, sell: 1 }, 10_000_000.0);
-        assert!(price.is_none());
+        // NOTE: Fully use liquidity provided by the first order. Note that it
+        // can send at most `100_000_000 / FEE_FACTOR` because of fees.
+        assert_approx_eq!(
+            pricegraph
+                .order_for_sell_amount(TokenPair { buy: 2, sell: 1 }, 1_000_000.0 * FEE_FACTOR)
+                .unwrap()
+                .buy,
+            100_000_000.0 / FEE_FACTOR
+        );
 
-        let price = pricegraph.estimate_limit_price(TokenPair { buy: 1, sell: 2 }, 1_000_000_000.0);
-        assert!(price.is_none());
+        // NOTE: For the next ~1_000_000 sell amount, we continue to fully use
+        // liquidity from the first order, and not include the second order
+        // because its limit price is too high.
+        assert_approx_eq!(
+            pricegraph
+                .order_for_sell_amount(
+                    TokenPair { buy: 2, sell: 1 },
+                    1_000_000.0 * FEE_FACTOR.powi(2),
```

Comment:
Is this assertion part of the "story"? I'm not sure I understand what it's supposed to tell me.
Maybe you multiply by an extra `FEE_FACTOR` here to nudge up the amount sold just a little and see that nothing changes?

Reply:
The `FEE_FACTOR^2` was that nudge :laughing:. I modified the test so that the "nudge" is more obvious.